core/src/conn.rs

```diff
@@ -1,10 +1,92 @@
+    Client,
+};
 use serde::{Deserialize, Serialize};
-use std::time::Duration;
+#[cfg(feature = "garbage_collect")]
+use std::sync::Weak;
+use std::{sync::Arc, time::Duration};
+use tokio::runtime::Runtime;
+
+// we make from UserOptions to Client and Weak<Runtime> so that we do not hold around
+// Clients and Runtimes that are no longer in use. In most cases it won't matter, but for drivers that live in
+// memory for a long time, this could be a problem.
+#[cfg(feature = "garbage_collect")]
+struct ClientMap(Vec<(UserOptions, Client, Weak<Runtime>)>);
+#[cfg(not(feature = "garbage_collect"))]
+struct ClientMap(Vec<(UserOptions, Client, Arc<Runtime>)>);
+
+impl ClientMap {
+    // new creates a new ClientMap.
+    fn new() -> Self {
+        ClientMap(Vec::new())
+    }
+
+    // get returns the client associated with the given user options if it exists.
+    #[cfg(feature = "garbage_collect")]
+    fn get(&mut self, user_options: &UserOptions) -> Option<(Client, Arc<Runtime>)> {
+        let mut to_remove = Vec::new();
+        for (i, (options, client, weak_rt)) in self.0.iter().enumerate() {
+            if options == user_options {
+                if let Some(rt) = weak_rt.upgrade() {
+                    return Some((client.clone(), rt));
+                // if somehow the Runtime cannot be upgraded, we want to remove this entry in the
+                // ClientMap, because the Runtime associated with the Client has been dropped,
+                // which will make the Topology hang forever.
+                } else {
+                    to_remove.push(i);
+                }
+            }
+        }
+        // We have to do this this way due to the borrow checker, it would be nice if we could just
+        // remove in the else above, but we cannot.
+        for i in to_remove.into_iter().rev() {
+            self.0.remove(i);
+        }
+        None
+    }
+
+    // get returns the client associated with the given user options if it exists.
+    #[cfg(not(feature = "garbage_collect"))]
+    fn get(&mut self, user_options: &UserOptions) -> Option<(Client, Arc<Runtime>)> {
+        for (options, client, rt) in self.0.iter() {
+            if options == user_options {
+                return Some((client.clone(), rt.clone()));
+            }
+        }
+        None
+    }
+
+    // gc is garbage collection of any clients that are no longer in use.
+    #[cfg(feature = "garbage_collect")]
+    fn gc(&mut self) {
+        self.0
+            .retain(|(_, _, weak_client)| Weak::strong_count(weak_client) != 0);
+    }
+
+    // insert inserts a new client into the client map keyed on the user options. Note this will
+    // insert duplicates and it is on the user to check if an entry already exists.
+    #[cfg(feature = "garbage_collect")]
+    fn insert(&mut self, user_options: UserOptions, client: &Client, rt: &Arc<Runtime>) {
+        self.0
+            .push((user_options, client.clone(), Arc::downgrade(rt)));
```

Comment:
In my implementation, which is still in progress, I have it keyed on the connection properties and plan to leave out things like log level, extJsonMode, since those are more driver related than connection.